### ISSUE TITLE
InvalidAuthorizationHeader raised after being verifyed

### ISSUE THREAD

**Issue by Pac0o** [ahopkins/sanic-jwt#93]
Hi 
I am working with sanic-jwt and marshmallow to make the requests data validation.

If data validation fails I am calling the abort method from sanic with the errors. This worked as intended but after I added the protected decorator to the method the result changed. 
Now I am receiving this response even when the endpoint is receiving a valid Authorization token.
```
{
    "exception": "InvalidAuthorizationHeader",
    "reasons": {
        "polygon": [
            "Missing data for required field."
        ]
    }
}
```
The reasons are the errors that I called the abort method with. 
If the validation is successful I dont get any error. 

One thing that worked to not get the InvalidAuthorizationHeader was to return a response with the authorization header instead of the abort methor. 

I am making something wrong? Or is this the intended behaviour??

Thank you!

**Comment by vltr** [comment]
At first, it looks like a problem inside your code, but let's dig deeper to
discover the root cause. Can you add `debug=True` to `Initialize` and see
if it produces a different output and/or a more verbose log?


Em ter, 10 de abr de 2018 20:15, Pac0o <notifications@github.com> escreveu:

**Comment by Pac0o** [comment]
I just did it and the result was the same, also on the console nothing new appeared. Tomorrow I'll will post more of my code to see if the problem can be replicated or someone can show where is my mistake. 

Thank you

**Comment by ahopkins** [comment]
@Pac0o There is a [customizable exception handler](http://sanic-jwt.readthedocs.io/en/latest/pages/endpoints.html#exception-handling) inside `sanic-jwt`. Indeed the response you are receiving the default response coming from it.

`InvalidAuthorizationHeader` is raised when there is not a prefix to the `Authentication` header. I think the most helpful thing to start debugging would be for you to send us what the request headers look like for a sample request.

It should look something like this:

    Authorization: Bearer <JWT>

The prefix `Bearer` is the default. You can [customize that](http://sanic-jwt.readthedocs.io/en/latest/pages/configuration.html#authorization-header-prefix) if needed.

If this is not the case, then my next question would be where are you doing the validation? Inside the request? In middleware? It does sound to me like the header is present, or else you would receive `MissingAuthorizationHeader`.

**Comment by vltr** [comment]
> `InvalidAuthorizationHeader` is raised when there is not a prefix to the `Authentication` header

I thought it might have something to do with `sanic-jwt`, but the message error is what bothers me:

```
"polygon": [
    "Missing data for required field."
]
```

I think this may be related to `mashmallow`. Here's an example of cURL when calling a `@protected` endpoint with and without `Bearer` prefix:

```
$ curl -H "Content-Type: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJleHAiOjE1MjM0NDc3MDYsImFwcF9uYW1lIjoidGVzdCJ9.D_aaI_SEAs0VDA6FXDoFRKeSsLTc2uU7sYqY2gA2B70" http://127.0.0.1:8000/protected
{"protected":true}
                                                                                                                     
$ curl -H "Content-Type: application/json" -H "Authorization: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJleHAiOjE1MjM0NDc3MDYsImFwcF9uYW1lIjoidGVzdCJ9.D_aaI_SEAs0VDA6FXDoFRKeSsLTc2uU7sYqY2gA2B70" http://127.0.0.1:8000/protected 
{"reasons":"Auth required.","exception":"Unauthorized"}
```

I know it sound strange to return `{"exception": "InvalidAuthorizationHeader", ...}`, but we'll need further details to see what's going on.

**Comment by Pac0o** [comment]
Hi, to minize the librarys in the middle I made a new project justi with sanic and sanic-jwt.

app.py
```
from sanic import Sanic
from sanic import Sanic
from sanic.response import json
from sanic_jwt.decorators import protected
from sanic.exceptions import abort

from sanic_jwt import Initialize

async def authenticate(request, *args, **kwargs):

    return {'user_id':1,'user':'test_user'}

app = Sanic()

Initialize(app,
            authenticate=authenticate,
            secret='Secret',
            debug=True)

@app.route("/test",methods=['POST'])
@protected()
async def test(request):
	abort(400,'this is a test')
    #return json({"hello": "world"})
```

curl request made with postman
```
curl -X POST \
  http://localhost:8005/test \
  -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1MjM0NjU1NTEsInVzZXJfaWQiOjF9.iRgSCvv5VImAG7xAmLB0mqvYwgRFColZ6APJcaBMhqc' \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: 07052d98-edb4-9869-c406-11f87e8b3c6b' \
  -d '
'
```

Status code: 400 (this is the code that I passed to the abort method and changes accordingly)
response
```

{
    "exception": "InvalidAuthorizationHeader",
    "reasons": "this is a test"
}
```

What I thought is that the abort method raises an exception that is cached by some method added by the protected decorator and there the Authorization header is validated again.

Let me know if I can help in any other way. 

Thank you!

**Comment by vltr** [comment]
Thanks, @Pac0o ! You're right, `abort` does raise an exception that is intercepted by the decorator, re-raising it as `InvalidAuthorizationHeader`. I personally think this is not the intended behavior, so I'll let @ahopkins add his thoughts in here too.

**Comment by ahopkins** [comment]
Foiled! My plan for nonsensical exception handling has been ruined! :rofl: 

In seriousness, I will come back and try and take a look later this evening. The code is pretty straightforward I believe on that one exception, so I am honestly not sure off the top of my head why it would be bleeding over like that. I'll have to take a rain check and get back to you.

**Comment by ahopkins** [comment]
I figured it out. Sort of a misunderstanding on my part, and perhaps lack of proper documentation on `sanic`.

Inside the `sanic-jwt` [exceptions definitions](https://github.com/ahopkins/sanic-jwt/blob/dev/sanic_jwt/exceptions.py), I followed the pattern used by [sanic exceptions](https://github.com/channelcat/sanic/blob/master/sanic/exceptions.py). Something like this:

```python
@add_status_code(401)
class AuthenticationFailed(SanicJWTException):
    ...
```

My assumption of that decorator was merely that it was setting the `status_code` for when that exception would be called. That assumption is incorrect now that I look more carefully at the [api reference](http://sanic.readthedocs.io/en/latest/sanic/api_reference.html?highlight=add_status_code#sanic.exceptions.add_status_code)

        Decorator used for adding exceptions to _sanic_exceptions.

And, more closely at what `_sanic_exceptions` actually is.

In short, you are receiving this because you called for a `400` and `InvalidAuthorizationHeader` was the last exception class to be declared as a `400`. Therefore, when you tell your app that there is a `400`, it goes to the `_sanic_exceptions` and does a `_sanic_exceptions.get(400)` which returns `InvalidAuthorizationHeader`.

The solution is to not use the decorator, and just assign the property to the class:

```python
class AuthenticationFailed(SanicJWTException):
    status_code = 400
    ...
```

Open and ripe for a PR if anyone wants to take a swing.
